go/vt/vtgate/planbuilder/insert.go

```diff
@@ -200,12 +185,150 @@ func buildInsertShardedPlan(ins *sqlparser.Insert, table *vindexes.Table) (engin
+	if len(ins.Columns) == 0 {
+		if !table.ColumnListAuthoritative {
+			return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "insert should contain column list or the table should have authoritative columns in vschema")
+		}
+		populateInsertColumnlist(ins, table)
+	}
+
+	// select plan will be taken as input to insert rows into the table.
+	plan, err := subquerySelectPlan(ins, vschema, reservedVars, true)
+	if err != nil {
+		return nil, err
+	}
+	eins.Input = plan
+
+	// auto-increment column is added explicility if not provided.
+	if err := modifyForAutoinc(ins, eins); err != nil {
+		return nil, err
+	}
+
+	// Fill out the 3-d Values structure
+	eins.VindexValueOffset, err = extractColVindexOffsets(ins, eins.ColVindexes)
+	if err != nil {
+		return nil, err
+	}
+
+	generateInsertSelectQuery(ins, eins)
+	return eins, nil
+}
+
+func subquerySelectPlan(ins *sqlparser.Insert, vschema plancontext.VSchema, reservedVars *sqlparser.ReservedVars, sharded bool) (engine.Primitive, error) {
+	selectStmt, queryPlanner, err := getStatementAndPlanner(ins, vschema)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate the columns to match on insert and select
+	// for sharded insert table only
+	if sharded {
+		if err := checkColumnCounts(ins, selectStmt); err != nil {
+			return nil, err
+		}
+	}
+
+	// Override the locking with `for update` to lock the rows for inserting the data.
+	selectStmt.SetLock(sqlparser.ForUpdateLock)
+
+	return queryPlanner(selectStmt, reservedVars, vschema)
+}
+
+func getStatementAndPlanner(
+	ins *sqlparser.Insert,
+	vschema plancontext.VSchema,
+) (selectStmt sqlparser.SelectStatement, configuredPlanner selectPlanner, err error) {
+	switch stmt := ins.Rows.(type) {
+	case *sqlparser.Select:
+		configuredPlanner, err = getConfiguredPlanner(vschema, buildSelectPlan, stmt, "")
+		selectStmt = stmt
+	case *sqlparser.Union:
+		configuredPlanner, err = getConfiguredPlanner(vschema, buildUnionPlan, stmt, "")
+		selectStmt = stmt
+	default:
+		err = vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: insert plan with %T", ins.Rows)
+	}
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return selectStmt, configuredPlanner, nil
+}
+
+func checkColumnCounts(ins *sqlparser.Insert, selectStmt sqlparser.SelectStatement) error {
+	if len(ins.Columns) < selectStmt.GetColumnCount() {
+		return vterrors.NewErrorf(vtrpcpb.Code_INVALID_ARGUMENT, vterrors.WrongValueCountOnRow, "Column count doesn't match value count at row 1")
+	}
+	if len(ins.Columns) > selectStmt.GetColumnCount() {
+		sel := sqlparser.GetFirstSelect(selectStmt)
+		var hasStarExpr bool
+		for _, sExpr := range sel.SelectExprs {
+			if _, hasStarExpr = sExpr.(*sqlparser.StarExpr); hasStarExpr {
+				break
+			}
+		}
+		if !hasStarExpr {
+			return vterrors.NewErrorf(vtrpcpb.Code_INVALID_ARGUMENT, vterrors.WrongValueCountOnRow, "Column count doesn't match value count at row 1")
+		}
+	}
```

Comment:
shouldn't we also check that the number of expanded select expressions (from the `StarExpr` in `selectStmt`) does not go over the number of columns in `ins`?

Reply:
that will happen at the runtime, i.e. in the engine, it cannot happen in the planner.